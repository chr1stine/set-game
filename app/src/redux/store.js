import reducer from './rootReducer';
import { createStore } from 'redux';

const store = createStore(reducer);

store.subscribe(() => {
  let state = store.getState();
  if (state.collectedSet) {
    store.dispatch({ type: 'set' });
  }
});

export default store;
