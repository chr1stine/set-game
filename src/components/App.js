import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import Board from './Board'

const App = ({deck})=>{
    // вытягивает из колоды следующие amount карт
    function pullNewCards(amount){
        let cards = []
        while(deck.cards.length > 0 && cards.length < amount){
            cards.push(deck.cards.pop())
        }

        if (deck.cards.length === 0){
            dispatch({type:'cards/finished'})
        }

        return cards
    }

    function handleStartGame(){
        dispatch({type:'game/started'})
    }

    function handlePullExtraCards(){
        dispatch({type:'game/extra'})
    }

    function handleHint(){
        dispatch({type:'hint',payload:cardsOnTable})
    }

    const dispatch = useDispatch()
    const takenCards = useSelector(state=>state.takenCards) // собранные в сет
    const selectedCards = useSelector(state=>state.selectedCards) // выделенные
    const gameFinished = useSelector(state=>state.gameFinished) // закончена ли игра
    const gameStarted = useSelector(state=>state.gameStarted) // начата ли игра
    const extraCards = useSelector(state=>state.extraCards)
    const cardsInitialized = useSelector(state=>state.cardsInitialized)
    const hint = useSelector(state=>state.hint);
    const cardsFinished = useSelector(state=>state.cardsFinished)
    const [cardsOnTable,setCardsOnTable] = useState([]) // выложенные на стол

    // console.log(`there are ${cardsOnTable.length} cards on table`)

    // первичное выкладывание 12 карт
    if (!cardsInitialized &&
        !gameFinished && 
         gameStarted){
        setCardsOnTable(pullNewCards(12))
        dispatch({type:'cards/initialized'})
    }

    // убирание правильного сета, если собран, и выкладывание новых карт по надобности
    if (takenCards.length > 0){
        let newCards = cardsOnTable.filter(c=>{
            return !_.includes(takenCards,c)
        })

        setCardsOnTable(newCards)

        dispatch({type:'set/taken'})
    }

    if (cardsInitialized && cardsOnTable.length < 12 + extraCards && !cardsFinished){
        setCardsOnTable([...cardsOnTable,...pullNewCards(3)])
    }

    if (gameFinished){
        alert('congrats! you won')
    }

    return (
        <div id="container">
            <Board extraCards={extraCards} cardsOnTable={cardsOnTable} 
                takenCards={takenCards} 
                selectedCards={selectedCards} hint={hint} />

            <button disabled={gameStarted} onClick={handleStartGame} id='start_game'>Начать игру</button>

            <button className={hint?.length == 0 && !cardsFinished ? 'hint' : ''} disabled={extraCards === 6 ||  !gameStarted} onClick={handlePullExtraCards} id='pull_extra_cards'>Выложить три карты</button>

            <button onClick={handleHint}>Подсказка</button>
        </div>
    )

}

export default App