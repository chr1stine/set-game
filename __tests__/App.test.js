import TestRenderer from 'react-test-renderer';
import App from '../src/components/App';
import { Provider } from 'react-redux';
import store from '../src/redux/store';

it('App renders', () => {
  const component = TestRenderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
