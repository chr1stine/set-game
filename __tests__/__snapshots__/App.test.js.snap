// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div
  id="container"
>
  <div
    id="buttons"
  >
    <button
      className="btn btn-outline-primary"
      disabled={false}
      id="start_game"
      onClick={[Function]}
    >
      Начать игру
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      id="pull_extra_cards"
      onClick={[Function]}
    >
      Выложить три карты
    </button>
    <button
      className="btn btn-outline-primary"
      disabled={true}
      onClick={[Function]}
    >
      Подсказка
    </button>
  </div>
</div>
`;
